File: app/src/main/cpp/wn-steam-client/src/pb/cauthentication.cpp
#include "wn_steam/pb/cauthentication.h"

#include <cstring>  // std::memcpy for float deserialization

#include "wn_steam/proto_wire.h"

namespace wn_steam::pb {

// ---------------------------------------------------------------------------
// CAuthentication_DeviceDetails
// ---------------------------------------------------------------------------

std::vector<uint8_t> CAuthentication_DeviceDetails::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);
    w.string_field(1, device_friendly_name);
    w.int32_field( 2, static_cast<int32_t>(platform_type));
    w.int32_field( 3, os_type);
    w.uint32_field(4, gaming_device_type);
    w.uint32_field(5, client_count);
    w.bytes_field( 6, std::span<const uint8_t>(machine_id.data(), machine_id.size()));
    return out;
}

// ---------------------------------------------------------------------------
// CAuthentication_GetPasswordRSAPublicKey
// ---------------------------------------------------------------------------

std::vector<uint8_t>
CAuthentication_GetPasswordRSAPublicKey_Request::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);
    w.string_field(1, account_name);
    return out;
}

std::optional<CAuthentication_GetPasswordRSAPublicKey_Response>
CAuthentication_GetPasswordRSAPublicKey_Response::deserialize(
    std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CAuthentication_GetPasswordRSAPublicKey_Response m;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (t->field_number) {
            case 1:
                if (auto v = r.string(); v) m.publickey_mod = std::move(*v);
                else return std::nullopt;
                break;
            case 2:
                if (auto v = r.string(); v) m.publickey_exp = std::move(*v);
                else return std::nullopt;
                break;
            case 3:
                if (auto v = r.u64(); v) m.timestamp = *v;
                else return std::nullopt;
                break;
            default:
                if (!r.skip(t->wire_type)) return std::nullopt;
                break;
        }
    }
    return m;
}

// ---------------------------------------------------------------------------
// CAuthentication_BeginAuthSessionViaCredentials_Request
// ---------------------------------------------------------------------------

std::vector<uint8_t>
CAuthentication_BeginAuthSessionViaCredentials_Request::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);

    // Field numbers verified against the canonical
    // steammessages_auth.steamclient.proto on SteamDatabase/Protobufs.
    // The earlier numbering (3↔4 off-by-one, 7↔8 SWAPPED) was wrong; Steam
    // silently returned a 7-byte stub response (just `interval=5.0` and
    // an empty `extended_error_message`) for every credentials attempt.
    //   1  device_friendly_name (legacy, not emitted)
    //   2  account_name
    //   3  encrypted_password
    //   4  encryption_timestamp
    //   5  remember_login (deprecated, not emitted)
    //   6  platform_type (legacy, not emitted — set inside device_details)
    //   7  persistence
    //   8  website_id
    //   9  device_details
    //   10 guard_data
    //   11 language
    //   12 qos_level
    w.string_field(2, account_name);
    w.string_field(3, encrypted_password);
    w.uint64_field(4, encryption_timestamp);
    w.int32_field( 7, static_cast<int32_t>(persistence));
    w.string_field(8, website_id);

    auto dd_body = device_details.serialize();
    if (!dd_body.empty()) {
        w.submessage_field(9, std::span<const uint8_t>(dd_body.data(), dd_body.size()));
    }

    w.string_field(10, guard_data);
    w.uint32_field(11, language);
    w.int32_field( 12, qos_level);
    return out;
}

// ---------------------------------------------------------------------------
// CAuthentication_AllowedConfirmation
// ---------------------------------------------------------------------------

std::optional<CAuthentication_AllowedConfirmation>
CAuthentication_AllowedConfirmation::deserialize(std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CAuthentication_AllowedConfirmation m;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (t->field_number) {
            case 1:
                if (auto v = r.i32(); v)
                    m.confirmation_type = static_cast<EAuthSessionGuardType>(*v);
                else return std::nullopt;
                break;
            case 2:
                if (auto v = r.string(); v) m.associated_message = std::move(*v);
                else return std::nullopt;
                break;
            default:
                if (!r.skip(t->wire_type)) return std::nullopt;
                break;
        }
    }
    return m;
}

// ---------------------------------------------------------------------------
// CAuthentication_BeginAuthSessionViaCredentials_Response
// ---------------------------------------------------------------------------

std::optional<CAuthentication_BeginAuthSessionViaCredentials_Response>
CAuthentication_BeginAuthSessionViaCredentials_Response::deserialize(
    std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CAuthentication_BeginAuthSessionViaCredentials_Response m;

    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (t->field_number) {
            case 1:
                if (auto v = r.u64(); v) m.client_id = *v; else return std::nullopt;
                break;
            case 2:
                if (auto v = r.bytes(); v) m.request_id.assign(v->begin(), v->end());
                else return std::nullopt;
                break;
            case 3:
                if (t->wire_type != proto::WireType::Fixed32) return std::nullopt;
                if (auto v = r.fixed32(); v) {
                    // Reinterpret 4 LE bytes as IEEE 754 float
                    float f;
                    uint32_t u = *v;
                    std::memcpy(&f, &u, sizeof(float));
                    m.interval = f;
                } else return std::nullopt;
                break;
            case 4: {
                auto sub = r.bytes();
                if (!sub) return std::nullopt;
                auto ac = CAuthentication_AllowedConfirmation::deserialize(*sub);
                if (!ac) return std::nullopt;
                m.allowed_confirmations.push_back(std::move(*ac));
                break;
            }
            case 5:
                if (auto v = r.u64(); v) m.steamid = *v; else return std::nullopt;
                break;
            case 6:
                if (auto v = r.string(); v) m.weak_token = std::move(*v);
                else return std::nullopt;
                break;
            case 7:
                if (auto v = r.string(); v) m.agreement_session_url = std::move(*v);
                else return std::nullopt;
                break;
            case 8:
                if (auto v = r.string(); v) m.extended_error_message = std::move(*v);
                else return std::nullopt;
                break;
            default:
                if (!r.skip(t->wire_type)) return std::nullopt;
                break;
        }
    }
    return m;
}

// ---------------------------------------------------------------------------
// CAuthentication_PollAuthSessionStatus
// ---------------------------------------------------------------------------

std::vector<uint8_t>
CAuthentication_PollAuthSessionStatus_Request::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);
    w.uint64_field(1, client_id);
    w.bytes_field( 2, std::span<const uint8_t>(request_id.data(), request_id.size()));
    if (token_to_revoke != 0) {
        w.tag(3, proto::WireType::Fixed64);
        for (int i = 0; i < 8; ++i) out.push_back(static_cast<uint8_t>(token_to_revoke >> (i * 8)));
    }
    return out;
}

std::optional<CAuthentication_PollAuthSessionStatus_Response>
CAuthentication_PollAuthSessionStatus_Response::deserialize(
    std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CAuthentication_PollAuthSessionStatus_Response m;

    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (t->field_number) {
            case 1:
                if (auto v = r.u64(); v) m.new_client_id = *v; else return std::nullopt;
                break;
            case 2:
                if (auto v = r.string(); v) m.new_challenge_url = std::move(*v);
                else return std::nullopt;
                break;
            case 3:
                if (auto v = r.string(); v) m.refresh_token = std::move(*v);
                else return std::nullopt;
                break;
            case 4:
                if (auto v = r.string(); v) m.access_token = std::move(*v);
                else return std::nullopt;
                break;
            case 5:
                if (auto v = r.boolean(); v) m.had_remote_interaction = *v;
                else return std::nullopt;
                break;
            case 6:
                if (auto v = r.string(); v) m.account_name = std::move(*v);
                else return std::nullopt;
                break;
            case 7:
                if (auto v = r.string(); v) m.new_guard_data = std::move(*v);
                else return std::nullopt;
                break;
            case 8:
                if (auto v = r.string(); v) m.agreement_session_url = std::move(*v);
                else return std::nullopt;
                break;
            default:
                if (!r.skip(t->wire_type)) return std::nullopt;
                break;
        }
    }
    return m;
}

// ---------------------------------------------------------------------------
// CAuthentication_UpdateAuthSessionWithSteamGuardCode_Request
// ---------------------------------------------------------------------------

std::vector<uint8_t>
CAuthentication_UpdateAuthSessionWithSteamGuardCode_Request::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);
    w.uint64_field(1, client_id);
    if (steamid != 0) {
        w.tag(2, proto::WireType::Fixed64);
        for (int i = 0; i < 8; ++i) out.push_back(static_cast<uint8_t>(steamid >> (i * 8)));
    }
    w.string_field(3, code);
    w.int32_field( 4, static_cast<int32_t>(code_type));
    return out;
}

// ---------------------------------------------------------------------------
// CAuthentication_BeginAuthSessionViaQR
// ---------------------------------------------------------------------------

std::vector<uint8_t>
CAuthentication_BeginAuthSessionViaQR_Request::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);
    w.string_field(1, device_friendly_name);
    w.int32_field( 2, static_cast<int32_t>(platform_type));
    auto dd_body = device_details.serialize();
    if (!dd_body.empty()) {
        w.submessage_field(3, std::span<const uint8_t>(dd_body.data(), dd_body.size()));
    }
    w.string_field(4, website_id);
    return out;
}

std::optional<CAuthentication_BeginAuthSessionViaQR_Response>
CAuthentication_BeginAuthSessionViaQR_Response::deserialize(
    std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CAuthentication_BeginAuthSessionViaQR_Response m;

    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (t->field_number) {
            case 1:
                if (auto v = r.u64(); v) m.client_id = *v; else return std::nullopt;
                break;
            case 2:
                if (auto v = r.string(); v) m.challenge_url = std::move(*v);
                else return std::nullopt;
                break;
            case 3:
                if (auto v = r.bytes(); v) m.request_id.assign(v->begin(), v->end());
                else return std::nullopt;
                break;
            case 4:
                if (t->wire_type != proto::WireType::Fixed32) return std::nullopt;
                if (auto v = r.fixed32(); v) {
                    float f;
                    uint32_t u = *v;
                    std::memcpy(&f, &u, sizeof(float));
                    m.interval = f;
                } else return std::nullopt;
                break;
            case 5: {
                auto sub = r.bytes();
                if (!sub) return std::nullopt;
                auto ac = CAuthentication_AllowedConfirmation::deserialize(*sub);
                if (!ac) return std::nullopt;
                m.allowed_confirmations.push_back(std::move(*ac));
                break;
            }
            case 6:
                if (auto v = r.i32(); v) m.version = *v; else return std::nullopt;
                break;
            default:
                if (!r.skip(t->wire_type)) return std::nullopt;
                break;
        }
    }
    return m;
}

}  // namespace wn_steam::pb

File: app/src/main/cpp/wn-steam-client/rust/src/pb/cauthentication.rs
use crate::proto_wire::{Reader, WireType, Writer};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[repr(i32)]
pub enum EAuthTokenPlatformType {
    Unknown = 0,
    SteamClient = 1,
    WebBrowser = 2,
    MobileApp = 3,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[repr(i32)]
pub enum EAuthSessionGuardType {
    Unknown = 0,
    None = 1,
    EmailCode = 2,
    DeviceCode = 3,
    DeviceConfirmation = 4,
    EmailConfirmation = 5,
    MachineToken = 6,
    LegacyMachineAuth = 7,
}

impl From<i32> for EAuthSessionGuardType {
    fn from(value: i32) -> Self {
        match value {
            1 => Self::None,
            2 => Self::EmailCode,
            3 => Self::DeviceCode,
            4 => Self::DeviceConfirmation,
            5 => Self::EmailConfirmation,
            6 => Self::MachineToken,
            7 => Self::LegacyMachineAuth,
            _ => Self::Unknown,
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[repr(i32)]
pub enum ESessionPersistence {
    Invalid = -1,
    Ephemeral = 0,
    Persistent = 1,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CAuthenticationDeviceDetails {
    pub device_friendly_name: String,
    pub platform_type: EAuthTokenPlatformType,
    pub os_type: i32,
    pub gaming_device_type: u32,
    pub client_count: u32,
    pub machine_id: Vec<u8>,
}

impl Default for CAuthenticationDeviceDetails {
    fn default() -> Self {
        Self {
            device_friendly_name: String::new(),
            platform_type: EAuthTokenPlatformType::SteamClient,
            os_type: 16,
            gaming_device_type: 0,
            client_count: 0,
            machine_id: Vec::new(),
        }
    }
}

impl CAuthenticationDeviceDetails {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        let mut w = Writer::new(&mut out);
        w.string_field(1, &self.device_friendly_name);
        w.int32_field(2, self.platform_type as i32);
        w.int32_field(3, self.os_type);
        w.uint32_field(4, self.gaming_device_type);
        w.uint32_field(5, self.client_count);
        w.bytes_field(6, &self.machine_id);
        out
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct GetPasswordRsaPublicKeyRequest {
    pub account_name: String,
}

impl GetPasswordRsaPublicKeyRequest {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        Writer::new(&mut out).string_field(1, &self.account_name);
        out
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct GetPasswordRsaPublicKeyResponse {
    pub publickey_mod: String,
    pub publickey_exp: String,
    pub timestamp: u64,
}

impl GetPasswordRsaPublicKeyResponse {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                1 => msg.publickey_mod = reader.string()?,
                2 => msg.publickey_exp = reader.string()?,
                3 => msg.timestamp = reader.u64()?,
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BeginAuthSessionViaCredentialsRequest {
    pub account_name: String,
    pub encrypted_password: String,
    pub encryption_timestamp: u64,
    pub website_id: String,
    pub persistence: ESessionPersistence,
    pub device_details: CAuthenticationDeviceDetails,
    pub guard_data: String,
    pub language: u32,
    pub qos_level: i32,
}

impl Default for BeginAuthSessionViaCredentialsRequest {
    fn default() -> Self {
        Self {
            account_name: String::new(),
            encrypted_password: String::new(),
            encryption_timestamp: 0,
            website_id: "Client".to_string(),
            persistence: ESessionPersistence::Persistent,
            device_details: CAuthenticationDeviceDetails::default(),
            guard_data: String::new(),
            language: 0,
            qos_level: 2,
        }
    }
}

impl BeginAuthSessionViaCredentialsRequest {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        let mut w = Writer::new(&mut out);
        w.string_field(2, &self.account_name);
        w.string_field(3, &self.encrypted_password);
        w.uint64_field(4, self.encryption_timestamp);
        w.int32_field(7, self.persistence as i32);
        w.string_field(8, &self.website_id);
        let dd = self.device_details.serialize();
        if !dd.is_empty() {
            w.submessage_field(9, &dd);
        }
        w.string_field(10, &self.guard_data);
        w.uint32_field(11, self.language);
        w.int32_field(12, self.qos_level);
        out
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AllowedConfirmation {
    pub confirmation_type: EAuthSessionGuardType,
    pub associated_message: String,
}

impl Default for AllowedConfirmation {
    fn default() -> Self {
        Self {
            confirmation_type: EAuthSessionGuardType::Unknown,
            associated_message: String::new(),
        }
    }
}

impl AllowedConfirmation {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                1 => msg.confirmation_type = EAuthSessionGuardType::from(reader.i32()?),
                2 => msg.associated_message = reader.string()?,
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct BeginAuthSessionViaCredentialsResponse {
    pub client_id: u64,
    pub request_id: Vec<u8>,
    pub interval: f32,
    pub allowed_confirmations: Vec<AllowedConfirmation>,
    pub steamid: u64,
    pub weak_token: String,
    pub agreement_session_url: String,
    pub extended_error_message: String,
}

impl Default for BeginAuthSessionViaCredentialsResponse {
    fn default() -> Self {
        Self {
            client_id: 0,
            request_id: Vec::new(),
            interval: 5.0,
            allowed_confirmations: Vec::new(),
            steamid: 0,
            weak_token: String::new(),
            agreement_session_url: String::new(),
            extended_error_message: String::new(),
        }
    }
}

impl BeginAuthSessionViaCredentialsResponse {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                1 => msg.client_id = reader.u64()?,
                2 => msg.request_id = reader.bytes()?.to_vec(),
                3 => {
                    if tag.wire_type != WireType::Fixed32 {
                        return None;
                    }
                    msg.interval = f32::from_bits(reader.fixed32()?);
                }
                4 => msg
                    .allowed_confirmations
                    .push(AllowedConfirmation::deserialize(reader.bytes()?)?),
                5 => msg.steamid = reader.u64()?,
                6 => msg.weak_token = reader.string()?,
                7 => msg.agreement_session_url = reader.string()?,
                8 => msg.extended_error_message = reader.string()?,
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct PollAuthSessionStatusRequest {
    pub client_id: u64,
    pub request_id: Vec<u8>,
    pub token_to_revoke: u64,
}

impl PollAuthSessionStatusRequest {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        let mut w = Writer::new(&mut out);
        w.uint64_field(1, self.client_id);
        w.bytes_field(2, &self.request_id);
        if self.token_to_revoke != 0 {
            w.tag(3, WireType::Fixed64);
            w.raw_bytes(&self.token_to_revoke.to_le_bytes());
        }
        out
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct PollAuthSessionStatusResponse {
    pub new_client_id: u64,
    pub new_challenge_url: String,
    pub refresh_token: String,
    pub access_token: String,
    pub had_remote_interaction: bool,
    pub account_name: String,
    pub new_guard_data: String,
    pub agreement_session_url: String,
}

impl PollAuthSessionStatusResponse {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                1 => msg.new_client_id = reader.u64()?,
                2 => msg.new_challenge_url = reader.string()?,
                3 => msg.refresh_token = reader.string()?,
                4 => msg.access_token = reader.string()?,
                5 => msg.had_remote_interaction = reader.boolean()?,
                6 => msg.account_name = reader.string()?,
                7 => msg.new_guard_data = reader.string()?,
                8 => msg.agreement_session_url = reader.string()?,
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct UpdateAuthSessionWithSteamGuardCodeRequest {
    pub client_id: u64,
    pub steamid: u64,
    pub code: String,
    pub code_type: EAuthSessionGuardType,
}

impl Default for UpdateAuthSessionWithSteamGuardCodeRequest {
    fn default() -> Self {
        Self {
            client_id: 0,
            steamid: 0,
            code: String::new(),
            code_type: EAuthSessionGuardType::Unknown,
        }
    }
}

impl UpdateAuthSessionWithSteamGuardCodeRequest {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        let mut w = Writer::new(&mut out);
        w.uint64_field(1, self.client_id);
        if self.steamid != 0 {
            w.tag(2, WireType::Fixed64);
            w.raw_bytes(&self.steamid.to_le_bytes());
        }
        w.string_field(3, &self.code);
        w.int32_field(4, self.code_type as i32);
        out
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BeginAuthSessionViaQrRequest {
    pub device_friendly_name: String,
    pub platform_type: EAuthTokenPlatformType,
    pub device_details: CAuthenticationDeviceDetails,
    pub website_id: String,
}

impl Default for BeginAuthSessionViaQrRequest {
    fn default() -> Self {
        Self {
            device_friendly_name: String::new(),
            platform_type: EAuthTokenPlatformType::MobileApp,
            device_details: CAuthenticationDeviceDetails::default(),
            website_id: "Mobile".to_string(),
        }
    }
}

impl BeginAuthSessionViaQrRequest {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        let mut w = Writer::new(&mut out);
        w.string_field(1, &self.device_friendly_name);
        w.int32_field(2, self.platform_type as i32);
        let dd = self.device_details.serialize();
        if !dd.is_empty() {
            w.submessage_field(3, &dd);
        }
        w.string_field(4, &self.website_id);
        out
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct BeginAuthSessionViaQrResponse {
    pub client_id: u64,
    pub challenge_url: String,
    pub request_id: Vec<u8>,
    pub interval: f32,
    pub allowed_confirmations: Vec<AllowedConfirmation>,
    pub version: i32,
}

impl Default for BeginAuthSessionViaQrResponse {
    fn default() -> Self {
        Self {
            client_id: 0,
            challenge_url: String::new(),
            request_id: Vec::new(),
            interval: 5.0,
            allowed_confirmations: Vec::new(),
            version: 0,
        }
    }
}

impl BeginAuthSessionViaQrResponse {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                1 => msg.client_id = reader.u64()?,
                2 => msg.challenge_url = reader.string()?,
                3 => msg.request_id = reader.bytes()?.to_vec(),
                4 => {
                    if tag.wire_type != WireType::Fixed32 {
                        return None;
                    }
                    msg.interval = f32::from_bits(reader.fixed32()?);
                }
                5 => msg
                    .allowed_confirmations
                    .push(AllowedConfirmation::deserialize(reader.bytes()?)?),
                6 => msg.version = reader.i32()?,
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[repr(i32)]
pub enum EAuthTokenRenewalType {
    None = 0,
    Allow = 1,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AccessTokenGenerateForAppRequest {
    pub refresh_token: String,
    pub steamid: u64,
    pub renewal_type: EAuthTokenRenewalType,
}

impl Default for AccessTokenGenerateForAppRequest {
    fn default() -> Self {
        Self {
            refresh_token: String::new(),
            steamid: 0,
            renewal_type: EAuthTokenRenewalType::None,
        }
    }
}

impl AccessTokenGenerateForAppRequest {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        let mut w = Writer::new(&mut out);
        w.string_field(1, &self.refresh_token);
        w.fixed64_field(2, self.steamid);
        w.int32_field(3, self.renewal_type as i32);
        out
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct AccessTokenGenerateForAppResponse {
    pub access_token: String,
    pub refresh_token: String,
}

impl AccessTokenGenerateForAppResponse {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                1 => msg.access_token = reader.string()?,
                2 => msg.refresh_token = reader.string()?,
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn credentials_request_uses_audited_field_numbers() {
        let msg = BeginAuthSessionViaCredentialsRequest {
            account_name: "user".to_string(),
            encrypted_password: "enc".to_string(),
            encryption_timestamp: 123,
            ..Default::default()
        };
        let bytes = msg.serialize();
        let fields = field_numbers(&bytes);
        assert!(fields.contains(&2));
        assert!(fields.contains(&3));
        assert!(fields.contains(&4));
        assert!(fields.contains(&7));
        assert!(fields.contains(&8));
        assert!(fields.contains(&9));
        assert!(!fields.contains(&6));
    }

    #[test]
    fn credentials_response_reads_float_interval_and_confirmations() {
        let mut confirmation = Vec::new();
        {
            let mut w = Writer::new(&mut confirmation);
            w.int32_field(1, EAuthSessionGuardType::DeviceCode as i32);
            w.string_field(2, "mobile");
        }
        let mut bytes = Vec::new();
        {
            let mut w = Writer::new(&mut bytes);
            w.uint64_field(1, 55);
            w.bytes_field(2, &[1, 2]);
            w.tag(3, WireType::Fixed32);
            w.raw_bytes(&2.5f32.to_bits().to_le_bytes());
            w.submessage_field(4, &confirmation);
            w.uint64_field(5, 765);
        }
        let msg = BeginAuthSessionViaCredentialsResponse::deserialize(&bytes).unwrap();
        assert_eq!(msg.client_id, 55);
        assert_eq!(msg.interval, 2.5);
        assert_eq!(
            msg.allowed_confirmations[0].confirmation_type,
            EAuthSessionGuardType::DeviceCode
        );
    }

    fn field_numbers(bytes: &[u8]) -> Vec<i32> {
        let mut reader = Reader::new(bytes);
        let mut out = Vec::new();
        while let Some(tag) = reader.next_tag() {
            out.push(tag.field_number);
            reader.skip(tag.wire_type);
        }
        out
    }
}
